src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java

```diff
@@ -0,0 +1,200 @@
+        // retry.
+        if (HttpStatus.CONFLICT.value() == e.getStatusCode()) {
+          logger.debug("Notebook instance {} already created.", instanceName.formatName());
+          return StepResult.getStepResultSuccess();
+        }
+        return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
+      }
+
+      creationOperation =
+          OperationUtils.pollUntilComplete(
+              creationOperation, Duration.ofSeconds(20), Duration.ofMinutes(12));
+      if (creationOperation.getOperation().getError() != null) {
+        throw new RetryException(
+            String.format(
+                "Error creating notebook instance {}. {}",
+                instanceName.formatName(),
+                creationOperation.getOperation().getError()));
+      }
+    } catch (IOException e) {
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  private static Instance createInstanceModel(FlightContext flightContext, String projectId) {
+    Instance instance = new Instance();
+    ApiGcpAiNotebookInstanceCreationParameters creationParameters =
+        flightContext
+            .getInputParameters()
+            .get(CREATE_NOTEBOOK_PARAMETERS, ApiGcpAiNotebookInstanceCreationParameters.class);
+    setFields(creationParameters, instance);
+
+    String serviceAccountEmail =
+        CreateServiceAccountStep.serviceAccountEmail(
+            flightContext.getWorkingMap().get(CREATE_NOTEBOOK_SERVICE_ACCOUNT_ID, String.class),
+            projectId);
+    instance.setServiceAccount(serviceAccountEmail);
+
+    // Create the AI Notebook instance in the service account proxy mode to control proxy access by
+    // means of IAM permissions on the service account.
+    // https://cloud.google.com/ai-platform/notebooks/docs/troubleshooting#opening_a_notebook_results_in_a_403_forbidden_error
+    ImmutableMap<String, String> metadata =
+        new ImmutableMap.Builder<String, String>().put("proxy-mode", "service_account").build();
+    instance.setMetadata(metadata);
+
+    setNetworks(instance, projectId, creationParameters.getLocation());
+
+    return instance;
+  }
+
+  private static void setFields(
+      ApiGcpAiNotebookInstanceCreationParameters creationParameters, Instance instance) {
+    instance
+        .setPostStartupScript(creationParameters.getPostStartupScript())
+        .setMachineType(creationParameters.getMachineType());
+    ApiGcpAiNotebookInstanceVmImage vmImageParameters = creationParameters.getVmImage();
+    if (vmImageParameters != null) {
+      instance.setVmImage(
+          new VmImage()
+              .setProject(vmImageParameters.getProjectId())
+              .setImageFamily(vmImageParameters.getImageFamily())
+              .setImageName(vmImageParameters.getImageName()));
+    }
+    ApiGcpAiNotebookInstanceContainerImage containerImageParameters =
```

Comment:
From the comments in the open_api, I got the idea the vm image and the container image were mutually exclusive. Is that true? This could would allow both or neither to be set.

Reply:
Yes. The GCP API will throw exceptions if both or neither are set. I was planning on doing creation parameter validation in the controller in a follow up PR.